### ISSUE TITLE
A new simple trick to mitigate the codebook collapse

### ISSUE THREAD

**Issue by BakerBunker** [lucidrains/vector-quantize-pytorch#170]
paper: https://arxiv.org/abs/2411.02038
repo: https://github.com/youngsheen/SimVQ

**Comment by lucidrains** [comment]
@BakerBunker have you tried the rotation trick from Fifty et al. yet?

**Comment by BakerBunker** [comment]
Not yet, just discovered this paper from arxiv. Maybe it can be used with rotation trick and further improve the codebook utilization.

**Comment by lucidrains** [comment]
@BakerBunker the rotation trick brought my code utilization up to 100% nearly instantaneously

**Comment by lucidrains** [comment]
@BakerBunker your opinion counts a lot as a citizen scientist, give it a try and let me know what you see!

**Comment by guillaumeguy** [comment]
I've been banging my head on the ResidualVQ collapse and, it appears that changing the optimizer from `AdamW` to `Adagrad` did the trick.

RE: rotation trick, it would be nice to add an example. Happy to take a stab

**Comment by lucidrains** [comment]
@guillaumeguy it is already added by the first author himself [here](https://github.com/lucidrains/vector-quantize-pytorch/pull/164)

**Comment by guillaumeguy** [comment]
Here are my numbers (the metric is the dot product between the original `x` and the `reconstructed_x` using 5 codes):

#### AdamW: Failed to converge in both cases (with low codebook utilization)
- 0.65 - No trick
- 0.65 - Rotation trick

#### AdaGrad
- 0.817 - No trick
- 0.821 - Rotation trick (slightly better)

**Comment by lucidrains** [comment]
@guillaumeguy thanks for sharing these results

what are you training on that only has 5 codes?  that seems unrealistic

**Comment by guillaumeguy** [comment]
I am creating [semantic tokens](https://arxiv.org/pdf/2305.05065). They use 3 codes in the paper FWIW.

**Comment by lucidrains** [comment]
@guillaumeguy oh I see.. they use different amounts of codes per depth. does the repo even support that?

**Comment by lucidrains** [comment]
let me read this TIGER paper in more detail tomorrow

**Comment by lucidrains** [comment]
@guillaumeguy have you heard of anyone successfully deploying something like TIGER?

**Comment by guillaumeguy** [comment]
Youtube said they had some success. Here is the [paper](https://arxiv.org/abs/2306.08121).

**Comment by lucidrains** [comment]
@guillaumeguy ok, I'll get out the ability for different codebook sizes tomorrow. think you need to test it out in that fashion

**Comment by lucidrains** [comment]
@guillaumeguy ok, this should be available now [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/master/tests/test_readme.py#L209)

**Comment by guillaumeguy** [comment]
Thanks @lucidrains ! Such a quick turnaround time.

Do you mind pointing me where in the paper you're seeing different codebook size? I'm reading 256 size x 3 codebooks: 

![image](https://github.com/user-attachments/assets/9ec82a03-c236-4933-90af-6849b1ef8f1e)

**Comment by lucidrains** [comment]
@guillaumeguy ah my bad, i misread that section.. ok, so the repo could support it by default. where were you getting 5 codes from?

**Comment by lucidrains** [comment]
@guillaumeguy think i'll keep the ability for different amount of codes across layers though, may come in handy

**Comment by guillaumeguy** [comment]
Mostly my experiment. The level 3 the paper chose is arbitrary and only capture the high level features of the input embeddings. I suspect that production systems need better granularity (Youtube went the other way by increasing the cardinality but I like the idea of just increase the # of codebooks). Will report back

**Comment by lucidrains** [comment]
@guillaumeguy oh i understand now, you are using 5 codebooks, each of size 256 codes?

**Comment by lucidrains** [comment]
@guillaumeguy ok, then i think your results are legit

**Comment by lucidrains** [comment]
@BakerBunker went through the paper and changed my mind, let me play around with it tomorrow morning and see if i can cleanly integrate it

**Comment by lucidrains** [comment]
ok, i've gone through the paper in detail. not 100% on this paper, but i think it is worth trying, and will probably redo this in a separate file. if all of a sudden everyone is clamoring over it, i can slowly port over working features from all the research of year's past in VQ

**Comment by lucidrains** [comment]
unfortunately, not seeing it for this paper

**Comment by youngsheen** [comment]
@lucidrains Hi, I'm the first author of the paper [SimVQ](https://github.com/youngsheen/SimVQ) and I appreciate your interest in our group’s work!

In our paper, **ALL** main results for both image and audio tasks are conducted with the codebook frozen, which is sampled from a Gaussian distribution. You can find the core code [here](https://github.com/youngsheen/SimVQ/blob/main/taming/modules/vqvae/quantize.py#L28-L33). We also conducted an ablation study with an unfrozen codebook in the paper (Trainable = Yes), where we observed slightly lower performance compared to the frozen variant. Notably, the unfrozen version can still achieve 100% codebook utilization.

<img width="804" alt="image" src="https://github.com/user-attachments/assets/8dfb1017-149a-4d96-ba00-eefacdac160e">

Please feel free to reach out if you have any questions. I would be grateful if you could consider adding our method to your excellent [vector-quantize-pytorch](https://github.com/lucidrains/vector-quantize-pytorch) repository.

**Comment by lucidrains** [comment]
@youngsheen hi Yongxin, thanks for reaching out and for the interesting paper

i tried out your formulation but it performed poorly. i could have made a mistake though, so let me retry. i noticed you add noise to the targets (std = `0.01`). where exactly is this noise applied? are you doing stochastic sampling?

**Comment by lucidrains** [comment]
@youngsheen do you see some overlap between your work and FSQ and LFQ? i see no mention of that in your paper, but i'm sure you thought about it

**Comment by youngsheen** [comment]
@lucidrains  Hi, thank you for trying out our formulation. Just to clarify, we only add noise to the targets in the toy examples to simulate small variations in the encoder outputs. This noise is not included in the main experiments. If you’d like, I’d be glad to help troubleshoot further—feel free to share your implementation details here, and we can take a look together.

**Comment by lucidrains** [comment]
@youngsheen ok thank you Yongxin. your offering to help is encouraging

i will circle back to this tomorrow and try it again

**Comment by lucidrains** [comment]
@youngsheen hi Yongxin, revisiting this this morning

do you see any issues [in this section of the code](https://github.com/lucidrains/vector-quantize-pytorch/commit/7959292f8ea1655eefe9dd0ecf99902be134b3df#diff-8650397c99e0e7ab7abb5e56fc847894cf9d45db2be9c15468a4eb94aa2d8379R80-R88)? that should capture the core idea right?

**Comment by youngsheen** [comment]
@lucidrains Hi, 
In vector_quantize_pytorch/sim_vq.py
Line 64: The standard deviation of the Gaussian distribution should be small, as is common in model initialization methods. For our experiments, we set std=e_dim**-0.5
Line 99-108: We did not apply normalization or the rotation trick. Instead, we directly used the STE in line 112.

Everything else looks good to me.

**Comment by lucidrains** [comment]
@youngsheen yes, i added the rotation trick because it was performing really poorly

let me try with the smaller init

**Comment by lucidrains** [comment]
@youngsheen no dice, the code utilization quickly collapses down to < 1 %

here is the [new init code](https://github.com/lucidrains/vector-quantize-pytorch/commit/007209dd7f9c3748ba2fa57faedb1377c48ab3bb#diff-8650397c99e0e7ab7abb5e56fc847894cf9d45db2be9c15468a4eb94aa2d8379R45)

**Comment by youngsheen** [comment]
<img width="518" alt="image" src="https://github.com/user-attachments/assets/72ab99ee-12fc-4521-86d1-4d62a6c6c69c">
it seems like you forgot to uncomment the STE code.

**Comment by lucidrains** [comment]
@youngsheen ah oops, ok yea it is uncommented

still the same performance

you can try this yourself by checking out the branch and running `pip install . && python examples/autoencoder_sim_vq.py`

**Comment by youngsheen** [comment]
@lucidrains Hi, I ran the code you provided and identified an issue with the commitment loss calculation. After correcting it, the codebook utilization quickly reaches nearly 100%.

In vector_quantize_pytorch/sim_vq.py, Line 73, commit_loss = 0.25 * ((x - quantized.detach()) ** 2).mean() + ((x.detach() - quantized) ** 2).mean()

Additionally, the current calculation of codebook utilization is incorrect. You should sum the activated codes over the training process or on the validation set for utilization tracking.


```python

sum_code = [0] * num_codes

def train(model, train_loader, train_iterations=1000, alpha=1.0):

    for _ in (pbar := trange(train_iterations)):

        for ind in indices.unique():
            sum_code[ind] = 1

        pbar.set_description(
            + f"active %: {sum(sum_code) / num_codes * 100:.3f}"
        )

```

**Comment by lucidrains** [comment]
@youngsheen hey, indeed it is a bit better, sitting around 30% code utilization (old method). it would also seem the `0.25` is crucial, as it suffers if i make it the same weight

i'm not sure i agree with your deeming a code is active across the entire training run. surely you need to have some time window?

**Comment by lucidrains** [comment]
@youngsheen in any case, it does seem to be an improvement over base VQ. let me play around with it some more

thank you for this exchange

**Comment by lucidrains** [comment]
can't believe it takes 7 years from Oord's initial paper until now for these solutions to emerge. have to play the long game with research

**Comment by lucidrains** [comment]
@BakerBunker you can try it out by invoking `from vector_quantize_pytorch import SimVQ`

i'll document this in the readme once i get out the residual flavor later this week

**Comment by cfifty** [comment]
Super cool work -- congrats @youngsheen! I bid quite strongly on this paper, but unfortunately, was not assigned to review it.

@lucidrains appreciate the tag and shoutouts to the rotation trick :) 

re:codebook utilization -- my understanding was that codebook utilization is computed on a single batch (i.e. # of codes used in this batch divided by total # of codes) and then this value is averaged across all batches. ViT-VQGAN has a good description, and this is how we computed codebook utilization in the rotation trick paper: 
<img width="568" alt="codebook_usage" src="https://github.com/user-attachments/assets/eeedbdec-77e5-4530-ae0d-33a13bc0d2af">

Recently, it seems like there's a more liberal interpretation. FSQ computes codebook utilization as "The fraction of the codewords that are used at least once when encoding the validation set."
<img width="590" alt="FSQ_usage" src="https://github.com/user-attachments/assets/2f28f52f-fc02-43fa-ab86-3c4479f98ce1">

My take-away from this is to be very clear which measure is used in the manuscript and highlight that different measures are being used across different works.

**Comment by lucidrains** [comment]
@cfifty ah i see! thanks for clearing up the two ways code utilization is being measured :pray:

**Comment by lucidrains** [comment]
@guillaumeguy ok, i got the residual flavor of sim vq built [here](https://github.com/lucidrains/vector-quantize-pytorch/commit/81d0f3b4c6fcfc5186e706fb817c1c182a179150#diff-d9cee953279014ae3c9536028cc2d6245448a0853870d4ab091568dbc6c91961R384). give it a test drive and let me know what you see on your end

**Comment by lucidrains** [comment]
@guillaumeguy also, i'm curious if you buy into the hand wavy arguments that residual VQ induces hierarchical reps in that recommendation paper

**Comment by guillaumeguy** [comment]
Thanks @lucidrains ! Why did the expected input dimension change? `ResidualVQ` was happy with `(B,dim)` but simVQ isn't.

```python
from vector_quantize_pytorch import ResidualSimVQ

B = 24
dim = 256
input = torch.randn(B,dim)

sim_vq = ResidualSimVQ(
            dim=dim,
            num_quantizers=4,  # specify number of quantizers
            codebook_size=256,
            rotation_trick=True,
        )

sim_vq(input)
```

Will get you: 
```
Equations:
    [c] d = 256 256
    b n = 24
Reason: Expansion '(2,)' of expression 'b n' does not match expansion '(1,)' of expression '24'
```

**Comment by lucidrains** [comment]
@guillaumeguy can you try the latest version?

**Comment by lucidrains** [comment]
@guillaumeguy seeing anything? 👀

**Comment by guillaumeguy** [comment]
it’s running as we speak (with performance slowly increasing). Should have some results tomorrow.

**Comment by guillaumeguy** [comment]
ok, the results are a bit disappointing (keeping everything else constant). 

|Method | Similarity (100=perfect) | active (%, perfect=100)|
|---|---|---|
|SimVQ | 82.3 | 4.0|
|ResidualVQ | 87.7 | 65.9|

I will try to tune the lr. This will probably help simVQ a bit

**Comment by lucidrains** [comment]
@guillaumeguy that doesn't seem too bad at all, 5% improvement, although code utilization for plain sim vq seems low

is this with Adam or adagrad? Curious if the collapse issue got mitigated or not

**Comment by lucidrains** [comment]
oh nvm, noticed you were comparing only with residual vq

both have rotation trick turned on?

**Comment by lucidrains** [comment]
you meant to write `ResidualSimVQ`, correct?

**Comment by lucidrains** [comment]
ok, I would say try the small MLP instead of single Linear that I have in the examples. Otherwise, if it still does not work, I'd recommend just sticking with vanilla residual VQ + rotation trick then

**Comment by lucidrains** [comment]
@guillaumeguy thank you for sharing this

**Comment by guillaumeguy** [comment]
it is using AdaGrad with lr=5e-3 and batch size of 1024. The encoder/decoder have 2 intermediary layers

**Comment by lucidrains** [comment]
@guillaumeguy I would be curious to see a sim vq + Adam experiment and see if it still collapses or not

that will be very telling

only if you have time or course!

**Comment by guillaumeguy** [comment]
Hold on ...  I used SimVQ instead of ResidualSimVQ. Let me try with `ResidualSimVQ`

**Comment by lucidrains** [comment]
how did you decide on using adagrad vs adam? was this from some paper or your own discovery?

**Comment by lucidrains** [comment]
> Hold on ... I used SimVQ instead of ResidualSimVQ. Let me try with `ResidualSimVQ`

Lol

**Comment by guillaumeguy** [comment]
> how did you decide on using adagrad vs adam? was this from some paper or your own discovery?

Own discovery.

**Comment by lucidrains** [comment]
> > how did you decide on using adagrad vs adam? was this from some paper or your own discovery?
> 
> Own discovery.

that's kinda cool! maybe worth a tweet

**Comment by guillaumeguy** [comment]
After a few iterations (~200 steps).  Looks like Ada works better and is less sensitive to learning rate.


|   description                                 |   lr     |   optimizer  |   Batch  |   eval_active (100=all)  |   eval_cmt_loss  |   eval_dot_prod (100=perfect)  |   eval_rec_loss  |   train_active  (100=all) |
|-----------------------------------------------|----------|--------------|----------|----------------|------------------|------------------|------------------|-----------------|
|   Ada vae 8 quantizers (dim 256)              |   0.005  |   ada        |   1024   |   52.25        |   0.00           |   75.00          |   0.02           |   55.71         |
|   SimVQ - Ada - vae 8 quantizers (dim 256)    |   0.005  |   ada        |   1024   |   0.59         |   0.00           |   67.56          |   0.02           |   0.59          |
|   SimVQ - Ada - vae 8 quantizers (dim 256)    |   0.4    |   ada        |   1024   |   0.78         |   9016951.00     |   18.86          |   0.03           |   0.93          |
|   Ada vae 8 quantizers (dim 256)              |   0.4    |   ada        |   1024   |   13.96        |   0.00           |   19.90          |   0.03           |   15.67         |
|   SimVQ - Adam - vae 8 quantizers (dim 256)   |   0.4    |   adam       |   1024   |   0.49         |   2223902208.00  |   19.69          |   0.03           |   0.39          |
|   adam vae 8 quantizers (dim 256)             |   0.4    |   adam       |   1024   |   0.39         |   0.00           |   18.27          |   0.03           |   0.39          |
|   adam vae 8 quantizers (dim 256)             |   0.005  |   adam       |   1024   |   49.76        |   0.00           |   65.03          |   0.02           |   62.40         |
|   SimVQ - Adam - vae 8 quantizers (dim 256)   |   0.005  |   Adam       |   1024   |   0.59         |   0.00           |   67.64          |   0.02           |   0.63          |


(SimVQ = ResidualSimVQ)

Both use the rotation trick. No shared codebook

**Comment by lucidrains** [comment]
thank you Guillaume, this is very helpful! I would say run one last experiment where you remove the rotation trick from SimVQ, just in case I broke it by removing one half of the commit loss

**Comment by guillaumeguy** [comment]
Without rotation trick: 
<img width="380" alt="image" src="https://github.com/user-attachments/assets/f5b708d1-4c36-4607-8025-fe9405f0b1f7">


RE: 200 steps, yes, the TIGER paper runs for 20k steps but I found it useful to run a lot more (it plateaued at around 150k)

**Comment by lucidrains** [comment]
aw ok, well I guess vanilla and rotation trick it is

**Comment by youngsheen** [comment]
Strange results—it seems like SimVQ is not working at all. I believe there must be an issue somewhere. Could you provide the code for me to test?

**Comment by ehsan-ami** [comment]
@youngsheen

**Comment by ehsan-ami** [comment]
@youngsheen Hi, congrats on the great work🎉!

I noticed a difference: should the bias for the codebook projection set to be false as in here? https://github.com/lucidrains/vector-quantize-pytorch/blob/3e4ce165774d3f5944f12ffb5ccd02848bb38df6/vector_quantize_pytorch/sim_vq.py#L62
In [your](https://github.com/youngsheen/SimVQ/blob/64c44c6319de17d80a3efdea23a7ea89465bf707/taming/modules/vqvae/quantize.py#L33) implementation you keep the default value which is bias=True

**Comment by lucidrains** [comment]
@ehsan-ami hey Ehsan! thanks for adding your voice

have you tried it yet? curious what your experience has been

**Comment by ehsan-ami** [comment]
@lucidrains Hi Phil! Thanks for your great repo!

Yes, I just did some experiments with both your implementation (w/ rotation trick) and Yongxin's implementation. 
With both of them I am observing is that codebook utilization starts from a low value and slowly increase up to some cap. value (3% utilization); after that it fluctuates . Here are my codebook utilization (entropy with the base of 2) plot:

The magenta color shows your implementation and the orange one is based on original implementation
![image](https://github.com/user-attachments/assets/174db16b-cd1c-495d-ac6e-97a64bcbbef3)
(I discontinued the orange experiment as it seemed that I am getting (almost) similar training stats)

SimVQ seems to be a promising method for training VQ-VAEs; though I am suspicious that it is sensitives to hyperparameter choice (optimizer, the ratio of embedding size to codebook size, codebook projection matrix initialization method, etc.).  At least for my model I couldn't make it to get good performance in terms of reconstruction loss and codebook utilization.

PS 1: I am using SimVQ for time series quantization 
Codebook size is 2048 (~=entropy 11), embedding_dim: 192, optimizer: AdamW with default betas
PS 2: I tried FSQ, which has no extra hyperparameter; it could get good utilization and reconstruction performance.
